refactor(docker): split dockerfile into multiple ones

varl opened the pull request:
## General idea

The `dhis2-core/Dockerfile` handles building DHIS2 from source, and then leaves the `dhis.war` available at `/dhis.war`, along with `CHECKSUM` files..

This base image can be tagged as necessary and be reused across multiple other assemblies at this point.

See [this comment](https://github.com/dhis2/dhis2-core/pull/3894#issuecomment-539416233) for container image build times.

## File structure

```
./dhis2-core
├── dhis-2
├── docker
│   ├── build-containers.sh
│   ├── extract-artifacts.sh
│   ├── publish-containers.sh
│   ├── README.md
│   ├── shared
│   │   ├── containers-list.sh
│   │   └── server.xml
│   ├── tomcat-alpine
│   │   ├── docker-entrypoint.sh
│   │   └── Dockerfile
│   └── tomcat-debian
│       ├── docker-entrypoint.sh
│       └── Dockerfile
├── Dockerfile
├── Jenkinsfile-publish-image
└── README.md
```

## Usage

### Build the core image (a.k.a. the base image)

```sh
docker build -t core:local .
```

### Extract artifacts from base image

The extract script accepts an image from which to extract known artifacts from as first positional parameter.

```sh
./docker/extract-artifacts.sh <image>
```

The script dumps the artifacts in `./docker/artifacts`:

```
./dhis2-core/docker/artifacts
├── dhis.war
├── sha256sum.txt
└── md5sum.txt
```

The `extract-artifacts` _always_ purges the `artifacts` directory to ensure consistency.

### Customize container images

Customize which containers will be built and published using the `shared/containers-list.sh` file: https://github.com/dhis2/dhis2-core/pull/3894/files#diff-553ea93c842ece0c53a6331741a73e37

### Build container images (Tomcat)

> ℹ️ Uses `containers-list.sh` to get a list of containers to build

```sh
./docker/build-containers.sh <image>
```

Given that `image` is `core:local` the result is:

```
REPOSITORY          TAG                           IMAGE ID            CREATED             SIZE
core                local-8.5.34-jre8-alpine      cc5bcfc1964c        4 seconds ago       372MB
core                local-8.0-jre8-slim           d9d8ad7d7352        18 seconds ago      506MB
core                local-8.5-jdk8-openjdk-slim   77979af6aaab        41 seconds ago      587MB
core                local-9.0-jdk8-openjdk-slim   35f234ab3688        59 seconds ago      588MB
core                local                         db4a67b2ca73        55 minutes ago      270MB
```

### Build custom WAR file into containers

The `build-containers` script will pick up and use any `dhis.war` from the `dhis2-core/docker/artifacts` directory by default. This can be used to create containers that contain a custom WAR-file.

> ⚠️ To use a custom `dhis.war` the checksum files must be removed (skips validation), or proper ones need to be generated.

E.g. how to build a custom `dhis.war` into the containers:

```sh
mv ~/my-custom-dhis.war ./docker/artifacts/dhis.war
./docker/build-containers.sh core:local-test
```

### Publish container images (Tomcat)

> ℹ️ Uses `containers-list.sh` to get a list of containers to publish

```sh
./docker/publish-containers.sh <image>
```

### Checksums

Checksums are now generated for the build artifact and stored alongside the WAR-file in sha256 and md5. These should generally be a part of our releases.

```
/ # ls -l /srv/dhis2
total 257984
...
-rw-r--r--    1 root     root     264106108 Oct  9 06:58 dhis.war
-rw-r--r--    1 root     root            44 Oct  9 06:58 md5sum.txt
-rw-r--r--    1 root     root            76 Oct  9 06:58 sha256sum.txt
...

/ # sha256sum -c /srv/dhis2/sha256sum.txt
/dhis.war: OK

/ # md5sum -c /srv/dhis2/md5sum.txt
/dhis.war: OK
```

### TODO

- [x] Supports Alpine and Debian images
- [x] Checksum generation 
- [x] Integrate into Jenkins pipeline
- [x] Naming convention on hub.docker.com/dhis2
- [x] Documentation
- [x] Script to publish containers

varl commented:
The structure is such that we can easily add a `docker/tomcat9/Dockerfile` as well and build that the same way.

Another option is to allow the `FROM` image in the Tomcat `Dockerfile` to be customized, though I prefer maintaining the Dockerfile separately. That way, we can still support the `tomcat:8.5.34-alpine` setup we have today in addition to the more official Tomcat Docker images based on Debian.

varl commented:
I found this structure easier to integrate into the `Jenkinsfile-publish-image` pipeline as well. This is where I ran into problems with `docker-compose` on my other branch.

This setup is very efficient at caching layers and the resulting images, the `docker-compose` was not, though I assume that's mostly because of my lack of knowledge of `docker-compose`.

varl commented:
Using a core image, building other Tomcat container images is quick

### Building an application server image (Tomcat **8.5.34-alpine** in this case):

```sh
varl@dhis2:~/dhis2-core$ time docker build -t dhis2-alpine:local . -f docker/tomcat/8.5.34-jre8-alpine/Dockerfile  --buil
d-arg image=core:local
Sending build context to Docker daemon  150.5MB
Step 1/11 : ARG image
Step 2/11 : FROM $image as build
 ---> 4902c81f6648
Step 3/11 : FROM tomcat:8.5.34-jre8-alpine
 ---> 4ac473a3dd92
Step 4/11 : ENV DHIS2_HOME=/DHIS2_home
 ---> Using cache
 ---> c029707bc383
Step 5/11 : COPY ./docker/tomcat/8.5.34-jre8-alpine/docker-entrypoint.sh /usr/local/bin/
 ---> 73ec12b11ac4
Step 6/11 : RUN rm -rf /usr/local/tomcat/webapps/* &&     mkdir /usr/local/tomcat/webapps/ROOT &&     chmod +rx /usr/local/bin/docker-entrypoint.sh &&     mkdir $DHIS2_HOME &&     addgroup -S tomcat &&     addgroup root tomcat &&     adduser -S -D -G tomcat tomcat
 ---> Running in 9a322823b4b5
Removing intermediate container 9a322823b4b5
 ---> 211b1d769381
Step 7/11 : RUN apk add --update --no-cache         bash          su-exec
 ---> Running in fcf6bec06353
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
(1/1) Installing su-exec (0.2-r0)
Executing busybox-1.28.4-r1.trigger
OK: 94 MiB in 63 packages
Removing intermediate container fcf6bec06353
 ---> 3d08354dccc3
Step 8/11 : COPY ./docker/shared/server.xml /usr/local/tomcat/conf
 ---> f5e44f347053
Step 9/11 : COPY --from=build /dhis.war /usr/local/tomcat/webapps/ROOT.war
 ---> 5b3d5dd1550d
Step 10/11 : ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]
 ---> Running in 1eb91b0760eb
Removing intermediate container 1eb91b0760eb
 ---> f10e4f9e2d2c
Step 11/11 : CMD ["catalina.sh", "run"]
 ---> Running in 12f66aea5c2f
Removing intermediate container 12f66aea5c2f
 ---> ba31599ec142
Successfully built ba31599ec142
Successfully tagged dhis2-alpine:local
```

#### Result

```
real    0m15.466s
user    0m0.423s
sys     0m0.338s
```

### Rebuilding in **Tomcat 8.5-jre8** container (cache demonstration):

```
varl@dhis2:~/dhis2-core$ time docker build -t dhis2-tomcat85:local . -f docker/tomcat/8.5-jdk8-openjdk-slim/Dockerfile  --b
uild-arg image=core2:local
Sending build context to Docker daemon  150.5MB
Step 1/11 : ARG image
Step 2/11 : FROM $image as build
 ---> 4902c81f6648
Step 3/11 : FROM tomcat:8.5-jdk8-openjdk-slim
8.5-jdk8-openjdk-slim: Pulling from library/tomcat
Digest: sha256:18495b9cfdded05bda9c93d1a1a50521197cb4a3b06e97a91be7d242fabf0bf0
Status: Downloaded newer image for tomcat:8.5-jdk8-openjdk-slim
 ---> 3fb5d3f06920
Step 4/11 : ENV DHIS2_HOME=/DHIS2_home
 ---> Using cache
 ---> f80e9f0eba31
Step 5/11 : COPY ./docker/shared/docker-entrypoint.sh /usr/local/bin/
 ---> Using cache
 ---> 55a86055b311
Step 6/11 : RUN rm -rf /usr/local/tomcat/webapps/* &&     mkdir /usr/local/tomcat/webapps/ROOT &&     chmod +rx /usr/local/bin/docker-entrypoint.sh &&     mkdir $DHIS2_HOME &&     adduser --system --disabled-password --group tomcat
 ---> Using cache
 ---> b7fd6c6c5c96
Step 7/11 : RUN apt-get update &&     apt-get install --no-install-recommends -y         util-linux         bash         unzip
 ---> Using cache
 ---> 0a67e6eb5ae2
Step 8/11 : COPY ./docker/shared/server.xml /usr/local/tomcat/conf
 ---> Using cache
 ---> b13e2c9bee7a
Step 9/11 : COPY --from=build /dhis.war /usr/local/tomcat/webapps/ROOT.war
 ---> Using cache
 ---> 3d9abb755425
Step 10/11 : ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]
 ---> Using cache
 ---> 8d6a0d5b7e60
Step 11/11 : CMD ["catalina.sh", "run"]
 ---> Using cache
 ---> ec2cfb23efc6
Successfully built ec2cfb23efc6
Successfully tagged dhis2-alpine:local
```

#### Result

```

real    0m5.476s
user    0m0.400s
sys     0m0.315s
```

amcgee commented:
@varl 👍 definitely better to use `docker` than `docker-compose`!  I think the multi-stage build approach should work well as a native solution.  Another option I just found is [habitus](https://www.habitus.io) which seems to do exactly what we want - multiple standalone stages, some with extractable artifacts, etc.  I haven't looking into the implementation there, but it seems a little more explicit than the multi-stage approach?  It might also allow the "runtime" image to be built with any .war file, even if it didn't come from a Docker core build image.

varl commented:
Using the `docker-build.sh` script that can be run from e.g. `dhis2-core/dhis-2/dhis2-e2e-tests` and still set the build context correctly, results in these images:

```
dhis2/core-dev               master                         63456a354860        22 seconds ago      270MB
dhis2/core-dev               master-8.0-jre8-slim           85e0fc404802        11 minutes ago      506MB
dhis2/core-dev               master-9.0-jdk8-openjdk-slim   5d7cbd2c3b44        12 minutes ago      588MB
dhis2/core-dev               master-8.5-jdk8-openjdk-slim   0b27e6f80ef7        53 minutes ago      587MB
```

varl commented:
For the `COPY` command to work the file must exist somewhere in the build context as a `Dockerfile` cannot traverse to the parent to grab files for security reasons.

If the `docker-build.sh` is not given a `WAR_FILE` name, it will assume a WAR-file must be created and build the core image.

If a `WAR_FILE` is given, it assumes it exists at `docker/artifact/${WAR_FILE}`. Though it can technically live anywhere within the build context that is currently set to the root repo.

varl commented on docker/core/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+RUN cp dhis-web/dhis-web-portal/target/dhis.war /dhis.war
+
+FROM alpine:latest
```
If we don't use a multi-stage build here, the result image is **5 GB**. Or we need to do some cleanup internally on the base image.

varl commented on docker/core/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN cp dhis-web/dhis-web-portal/target/dhis.war /dhis.war
+
+FROM alpine:latest
+COPY --from=build /dhis.war /dhis.war
```
If we don't use a multi-stage build here, the result image is **5 GB**. Or we need to do some cleanup internally on the base image.

varl commented:
Having written the script that builds the core and/or tomcat images, I have a sneaking suspicion that we should not try to keep the `docker/core` and `docker/tomcat-*` `Dockerfile`s in the same place.

The `docker/core/Dockerfile` is only about building the core, and the `docker/tomcat-*/Dockerfile`s are about how to deploy it.

:thinking:

varl commented:
@amcgee I took a look at Habitus and seems pretty neat. We still have to do the splitting of Docker-images in this PR, though Habitus looks like it can replace eventual build scripts with their runtime.

Further investigation is needed on how Habitus works with Jenkins, how it affects build times, etc.

varl commented:
Commit 34d8f9f is a fun one!

I've changed the way we would publish the DHIS2 "base image" (the image that only contains the dhis.war and checksums) to hub.docker.com/u/dhis2:

When building a tag we suffix `-release`, and when we build a branch we suffix `-snapshot`. Before we used tags to publish to target `dhis2/core` and `dhis2/core-dev` respectively so the check is as good as it was.

```
dhis2/base:2.34.0-release
dhis2/base:master-snapshot
dhis2/base:2.33-snapshot
```

The alternative if we want to mirror the existing structure:

```

dhis2/base:2.34.0
dhis2/base-dev:master
dhis2/base-dev:2.33
```

Either way, this gives us a new place under Docker Hub to host our "base images" separately from the complete, deployment ready, images that we have already.

varl commented:
> 
> 
> I suppose this is WIP? This commit basically skips publishing the image with full environment to core, core-dev channels.

It is, did you see this comment: https://github.com/dhis2/dhis2-core/pull/3894#issuecomment-539982395?

vilkg commented:
> > I suppose this is WIP? This commit basically skips publishing the image with full environment to core, core-dev channels.
> 
> It is, did you see this comment: [#3894 (comment)](https://github.com/dhis2/dhis2-core/pull/3894#issuecomment-539982395)?

Yeah, but I don't get why. And how will the API tests run.

varl commented:
The general idea is to separate the "base image" from the "deployment image" (lack of better terms).

Maybe it's easier align with the current repository names to demonstrate the idea:

**Pipeline 1: build and publish the base image**
- \
  Jenkins builds a commit on `master` into a base image: `dhis2/base-dev:master`
- \
  If the build passes, it gets published to `dhis2/base-dev:master`.

**Pipeline 2: test and publish the container images**
- \
  Jenkins pulls the built image to assemble the containers we want, e.g.
  - `dhis2/core-dev:master-tomcat-8.5-jdk8-openjdk-slim`
  - `dhis2/core-dev:master-tomcat-8.5.34-jre8-alpine`
- \
  Jenkins executes the API tests against all of the containers to verify that they are functional (this would probably take too long so maybe just the default one, but ultimately we do want to run tests against all official releases on the dhis2/core channels)
- \
  Once the tests pass we can extract the WAR-file from the image that we can guarantee was used through the entire chain and upload that to S3 for distribution.
- \
  At this point we can publish all of the containers to hub.docker.com/u/dhis2/core-dev as well.

### Easier to test new container images

A benefit of keeping the "base image" around is that if we want to create different constellations container images, you can use the exact same base image and only vary the container image.

Ok, we can do all of that by hand today with a few drawbacks. Doing so we lose guarantees across the board by downloading and copying around artifacts unless we do a parity check on it between each step, which is time consuming and annoying.

### Container images are verified using API tests, base images are not

The benefit of splitting the "base images" (dhis2/base) and "container images" (dhis2/core) is that we can allow tests to run asynchronously against dhis2/core images (with tomcat) and _only_ publish those if they pass.

For the base images, we would allow them to be potentially broken until they get bundled into a container.

We can run pipeline 1 a lot faster than pipeline 2, and since they would be decoupled, we can run pipeline 2 less often than pipeline 1 in a separate job.

Now, I'm not sure exactly how our Jenkins is setup and if that's feasible or not, but since this is a WIP I figured I'd mock out one candidate implementation for how it could work in theory.

vilkg commented:
> Agreed, having a linear pipeline is preferable. We can run everything sequentially in this PR; that means running the API-tests against 4 Tomcat environments instead of 1.
> I expect that to take around 20 minutes instead of 5 minutes.

We can easily parallelise the stages within the pipeline, so if we "add horsepower", should take the same time. However, we don't necessarily have to run the tests against all versions right away, we can test the default and assume others are working, until we are proven wrong. 

> Splitting into "base images" and "container images" allows that flexibility.
I am ALL FOR the split!! I just don't like the split in the pipeline itself (having separate jobs for building war and building core env.

varl commented:
> We can easily parallelise the stages within the pipeline, so if we "add horsepower", should take the same time. 

Cool! ⚡️ 

> However, we don't necessarily have to run the tests against all tomcat versions right away, we can test the default and assume others are working, until we are proven wrong.

Alright, I've adjusted the scripts. Is it possible to test the Jenkins Pipeline locally somehow? Would be interested to see how that works.

---

## PR updates!

`build-containers.sh` produces the structure we have today. Note that `dhis2/core-dev:master` is available and matches the name in: https://hub.docker.com/r/dhis2/core-dev/tags

```
REPOSITORY          TAG                                  IMAGE ID            CREATED             SIZE
dhis2/core-dev      master-tomcat-8.5.34-jre8-alpine      8c697afdb4d1        18 seconds ago      372MB
dhis2/core-dev      master-tomcat-8.0-jre8-slim           ecab295c9003        27 seconds ago      506MB
dhis2/core-dev      master-tomcat-9.0-jdk8-openjdk-slim   9014116b0cde        42 seconds ago      588MB
dhis2/core-dev      master                                50ebde485813        56 seconds ago      587MB
dhis2/core-dev      master-tomcat-8.5-jdk8-openjdk-slim   50ebde485813        56 seconds ago      587MB
```

In addition we have many other containers to test if we want to [in the future].

I updated the Jenkins Pipeline in a57f972 (after reverting 34d8f9f) so it first builds the "base image", then the "container image", then runs the API tests against the default Tomcat (defined in a variable in `build-containers.sh` for now) as it normally would.

When everything is OK it publishes the base image and container image. There's a `TODO` to have it deploy all container images but I think we need another script for that.

varl commented:
>I don't think we agreed on separate pipelines, just that we should have the war as well to allow people building their own images?

You're right that we didn't settle on different pipelines. I meant that we agreed that "building our software" and "bundling our software" should be handled separately.

They can be handled separately in the same pipeline and I went about it with separate pipelines in mind. I've addressed it to be "separate in the same pipeline" and inline with this part of your comment above:

>However, we don't necessarily have to run the tests against all tomcat versions right away, we can test the default and assume others are working, until we are proven wrong.

Thanks for reviewing! 😸👍

varl commented:
Rebased on `master` and ready for review @vilkg @amcgee @Philip-Larsen-Donnelly.

varl commented on Dockerfile:
```diff
@@ -1,80 +1,38 @@
+FROM alpine:latest
+COPY --from=build /dhis.war /dhis.war
+COPY --from=build /sha256sum.txt /sha256sum.txt
+COPY --from=build /md5sum.txt /md5sum.txt
```
Maybe we should put the artifacts under `/srv/dhis2`?

http://www.tldp.org/LDP/Linux-Filesystem-Hierarchy/html/srv.html

varl commented:
Thanks, let's see if this bird flies.
